src/lasp_divergence_client.erl

```diff
@@ -0,0 +1,207 @@
+handle_cast(Msg, State) ->
+    lager:warning("Unhandled messages: ~p", [Msg]),
+    {noreply, State}.
+
+%% @private
+-spec handle_info(term(), #state{}) -> {noreply, #state{}}.
+handle_info(event, #state{actor=Actor, events=Events0}=State) ->
+    lasp_marathon_simulations:log_message_queue_size("event"),
+
+    LocalEvents = case lasp_workflow:is_task_completed(convergence, 1) of
+        true ->
+            Events1 = Events0 + 1,
+
+            Element = atom_to_list(Actor),
+
+            perform_update(Element, Actor, Events1),
+
+            case max_events_reached(Events1) of
+                true ->
+                    lager:info("All events done. Node: ~p", [Actor]),
+
+                    %% Update Simulation Status Instance
+                    lasp_workflow:task_completed(events, node()),
+                    log_convergence(),
+                    schedule_check_simulation_end();
+                false ->
+                    schedule_event()
+            end,
+            Events1;
+        false ->
+            schedule_event(),
+            Events0
+    end,
+
+    {noreply, State#state{events=LocalEvents}};
+
+handle_info(check_simulation_end, #state{actor=Actor}=State) ->
+    lasp_marathon_simulations:log_message_queue_size("check_simulation_end"),
+
+    case lasp_workflow:is_task_completed(events) of
+        true ->
+            lager:info("Performing final anti-entropy pass."),
+            ObjectFilterFun = fun(_) -> true end,
+            lasp_distribution_backend:blocking_sync(ObjectFilterFun),
+
+            lager:info("All nodes did all events. Node ~p", [Actor]),
+            lasp_instrumentation:stop(),
+            lasp_support:push_logs(),
+            lasp_workflow:task_completed(logs, node());
+        false ->
+            schedule_check_simulation_end()
+    end,
+
+    {noreply, State};
+
+handle_info(Msg, State) ->
+    lager:warning("Unhandled messages: ~p", [Msg]),
+    {noreply, State}.
+
+%% @private
+-spec terminate(term(), #state{}) -> term().
+terminate(_Reason, _State) ->
+    ok.
+
+%% @private
+-spec code_change(term() | {down, term()}, #state{}, term()) -> {ok, #state{}}.
+code_change(_OldVsn, State, _Extra) ->
+    {ok, State}.
+
+%%%===================================================================
+%%% Internal functions
+%%%===================================================================
+
+%% @private
+schedule_event() ->
+    timer:send_after(?EVENT_INTERVAL, event).
+
+%% @private
+schedule_check_simulation_end() ->
+    timer:send_after(?STATUS_INTERVAL, check_simulation_end).
+
+%% @private
+log_convergence() ->
+    case lasp_config:get(instrumentation, false) of
+        true ->
+            lasp_instrumentation:convergence();
+        false ->
+            ok
+    end.
+
+%% @private
+max_events() ->
+    lasp_config:get(max_events, ?MAX_EVENTS_DEFAULT).
+
+%% @private
+max_events_reached(Events) ->
+    max_events() == Events orelse lasp_config:get(events_generated, false).
```

Comment:
`lasp_config:get(events_generated, false)`
should be enough, assuming the enforce function will run